mcp_foundry_knowledge/data_access_objects/dao.py

```diff
@@ -0,0 +1,517 @@
+from azure.search.documents.indexes.models import SearchIndex, SearchIndexer, SearchIndexerDataSourceConnection
+
+
+class SearchBaseDao:
+    """
+    Base class for Azure Cognitive Search data access operations.
+
+    Handles environment configuration and authentication setup
+    for interacting with Azure AI Search services.
+    """
+
+    def __init__(self):
+        """
+        Initializes the SearchBaseDao by reading configuration from environment variables.
+        """
+        self.authentication_method = self._get_env_variable("AZURE_AUTHENTICATION_METHOD", "api-search-key")
+        self.service_endpoint = self._get_env_variable("AZURE_AI_SEARCH_ENDPOINT")
+        self.api_version = self._get_env_variable('AZURE_AI_SEARCH_API_VERSION', '2025-03-01-preview')
+
+    @staticmethod
+    def _get_env_variable(key: str, default_value: str | None = None) -> str:
+        """
+        Retrieves an environment variable value or returns a default if not set.
+
+        Args:
+            key (str): The name of the environment variable.
+            default_value (str | None): Optional fallback value.
+
+        Returns:
+            str: The value of the environment variable or the default.
+        """
+        return os.environ.get(key, default_value)
+
+    def _fetch_credentials(self) -> AzureKeyCredential | DefaultAzureCredential:
+        """
+        Fetches the appropriate credentials for Azure Search based on the configured authentication method.
+
+        Returns:
+            AzureKeyCredential | DefaultAzureCredential: A credential object for authenticating requests.
+
+        Raises:
+            Exception: If the authentication method is missing or invalid.
+        """
+        if self.authentication_method == 'api-search-key':
+            api_key = self._get_env_variable('AZURE_AI_SEARCH_API_KEY')
+            credential = AzureKeyCredential(api_key)
+            return credential
+        elif self.authentication_method == 'service-principal':
+            credential = DefaultAzureCredential()
+            return credential
+
+        error_message = (
+            "AZURE_AUTHENTICATION_METHOD was not specified or is invalid. "
+            "Must be one of api-search-key or service-principal"
+        )
+        raise Exception(error_message)
+
+
+class SearchIndexDao(SearchBaseDao):
+    """
+    Data Access Object (DAO) for interacting with Azure AI Search Indexes.
+
+    Inherits configuration and authentication from SearchBaseDao.
+    """
+
+    def __init__(self):
+        """
+        Initializes the SearchIndexDao with a SearchIndexClient instance.
+        """
+        super().__init__()
+        credentials = self._fetch_credentials()
+        self.client = SearchIndexClient(self.service_endpoint, credentials, api_version=self.api_version)
+
+    def close(self):
+        """Shuts down the Data Access Object instance and associated resources
+
+        :rtype: None
+        """
+        self.client.close()
```

Comment:
Is client close required? Should we close the client in every tool?

Reply:
After running it a couple of times, I noticed that some tool calls were not going back to the same index each time so we need to close the connection to free resources. That is why it is closing the connections after the query is done. I dont think there is any benefit to leaving the connections to AI Search Service open